core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3380,14 +3380,18 @@ public FileSystemCommand workerHeartbeat(long workerId, List<Long> persistedFile
       Preconditions.checkArgument(((InodeFile) inode).isCompleted(),
           PreconditionMessage.FILE_TO_PERSIST_MUST_BE_COMPLETE);
       InodeFile file = (InodeFile) inode;
-      // TODO(manugoyal) figure out valid behavior in the un-persist case
-      Preconditions
-          .checkArgument(options.getPersisted(), PreconditionMessage.ERR_SET_STATE_UNPERSIST);
-      if (!file.isPersisted()) {
-        file.setPersistenceState(PersistenceState.PERSISTED);
-        persistedInodes = propagatePersistedInternal(inodePath, false);
-        file.setLastModificationTimeMs(opTimeMs);
-        Metrics.FILES_PERSISTED.inc();
+      if (options.getPersisted()) {
+        if (!file.isPersisted()) {
+          file.setPersistenceState(PersistenceState.PERSISTED);
+          persistedInodes = propagatePersistedInternal(inodePath, false);
+          file.setLastModificationTimeMs(opTimeMs);
+          Metrics.FILES_PERSISTED.inc();
+        }
+      } else {
+        // Make sure we don't un-persist file once it is persisted
+        Preconditions.checkArgument(file.getPersistenceState() == PersistenceState.TO_BE_PERSISTED,
```

Comment:
Should this be `!= PERSISTED` ? What if the state was already `NOT_PERSISTED`. Should that be allowed?

Reply:
We always lock the inode before check & update the persisted state so such race condition should never happen. I can see that we might want to be tolerant to overwriting the same state in case of any future cases. There is also another state `PersistenceState.LOST`, would it be a valid transition from that to `PersistenceState.NOT_PERSISTED`?